packages/ampd-sdk/src/event/event_handler.rs

```diff
@@ -169,30 +138,32 @@ where
     async fn process_stream(
         &self,
         element: error_stack::Result<Event, Error>,
+        client: &mut impl client::Client,
         token: &CancellationToken,
-    ) -> HandlerTaskAction {
-        let action = self
-            .parse_event(element)
-            .and_then(|event| self.handle_event(event, token))
-            .and_then(|msgs| self.broadcast_msgs(msgs, token))
-            .await
-            .map(|_| HandlerTaskAction::Continue);
-
-        action.unwrap_or_else(|err_action| err_action)
+    ) {
+        if let Some(event) = self.parse_event(element).await {
+            if let Some(msgs) = self.handle_event(event, token).await {
+                self.broadcast_msgs(client, msgs, token).await;
+            }
+        }
     }
 
-    async fn parse_event(
-        &self,
-        element: error_stack::Result<Event, Error>,
-    ) -> Result<H::Event, HandlerTaskAction> {
+    #[instrument]
+    async fn parse_event(&self, element: error_stack::Result<Event, Error>) -> Option<H::Event> {
         element
             .inspect(Self::log_block_boundary)
             .and_then(|event| {
                 H::Event::try_from(event.clone())
                     .change_context(Error::EventConversion)
                     .attach_printable(json!({ "event": event }))
             })
-            .map_err(self.event_subscription_error_cb)
+            .map_err(|err| {
+                error!(
+                    err = report::LoggableError::from(&err).as_value(),
+                    "failed to parse events"
```

Comment:
that's actually not fully correct. you end up in this case if the element is Err from the get go. I believe it would be better to deal with the Err case already in process_stream and then have `async fn parse_event(&self, element: Event) -> Option<H::Event> {